File: src/main/java/com/ticketing/event/model/Event.java
package com.ticketing.event.model;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

import com.ticketing.booking.model.TicketTier;
import com.ticketing.user.model.User;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import jakarta.persistence.Version;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "events")
public class Event {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "organizer_id", nullable = false)
    private User organizer;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "venue_id", nullable = false)
    private Venue venue;

    @Column(name = "cover_image_url")
    private String coverImageUrl;

    @Column(name = "start_date", nullable = false)
    private Instant startDate;

    @Column(name = "end_date", nullable = false)
    private Instant endDate;

    @Column(name = "sales_open_date")
    private Instant salesOpenDate;

    @Column(name = "sales_close_date")
    private Instant salesCloseDate;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private EventStatus status;

    @Column(name = "dynamic_pricing_enabled", nullable = false)
    private Boolean dynamicPricingEnabled;

    @Column(name = "waitlist_enabled", nullable = false)
    private Boolean waitlistEnabled;

    @Version
    @Column(nullable = false)
    private Long version;

    @Column(name = "created_at", nullable = false)
    private Instant createdAt;

    @Column(name = "updated_at")
    private Instant updatedAt;

    @OneToMany(mappedBy = "event", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<TicketTier> ticketTiers = new ArrayList<>();

    @PrePersist
    void onCreate() {
        Instant now = Instant.now();
        createdAt = now;
        updatedAt = now;
        if (status == null) {
            status = EventStatus.DRAFT;
        }
        if (dynamicPricingEnabled == null) {
            dynamicPricingEnabled = Boolean.FALSE;
        }
        if (waitlistEnabled == null) {
            waitlistEnabled = Boolean.FALSE;
        }
    }

    @PreUpdate
    void onUpdate() {
        updatedAt = Instant.now();
    }
}

File: frontend/src/types/event.ts
export type EventStatus =
  | "DRAFT"
  | "PUBLISHED"
  | "SALES_OPEN"
  | "SALES_CLOSED"
  | "COMPLETED"
  | "ARCHIVED";

export type OrganizerInfo = {
  id: number;
  name: string;
  email: string;
};

export type TicketTierResponse = {
  id: number;
  name: string;
  price: number;
  availableQuantity: number;
  description: string | null;
};

export type CategoryResponse = {
  id: number;
  name: string;
  description: string | null;
  iconUrl: string | null;
};

export type VenueResponse = {
  id: number;
  name: string;
  address: string;
  city: string;
  country: string;
  totalCapacity: number;
};

export type EventResponse = {
  id: number;
  title: string;
  description: string | null;
  organizerId: number | null;
  categoryId: number | null;
  venueId: number | null;
  startDate: string;
  endDate: string;
  salesOpenDate: string | null;
  salesCloseDate: string | null;
  coverImageUrl: string | null;
  status: EventStatus;
  dynamicPricingEnabled: boolean;
  waitlistEnabled: boolean;
};

export type EventDetailResponse = Omit<EventResponse, "organizerId" | "categoryId" | "venueId"> & {
  organizer: OrganizerInfo;
  category: CategoryResponse;
  venue: VenueResponse;
  ticketTiers: TicketTierResponse[];
};

export type EventFilters = {
  q?: string;
  city?: string;
  categoryId?: number;
};
